x-pack/plugins/apm/server/lib/connections/get_connection_metrics/get_destination_map.ts

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { QueryDslQueryContainer } from '@elastic/elasticsearch/api/types';
+import objectHash from 'object-hash';
+import { getOffsetInMs } from '../../../../common/utils/get_offset_in_ms';
+import { ENVIRONMENT_NOT_DEFINED } from '../../../../common/environment_filter_values';
+import { asMutableArray } from '../../../../common/utils/as_mutable_array';
+import { AgentName } from '../../../../typings/es_schemas/ui/fields/agent';
+import {
+  AGENT_NAME,
+  EVENT_OUTCOME,
+  PARENT_ID,
+  SERVICE_ENVIRONMENT,
+  SERVICE_NAME,
+  SPAN_DESTINATION_SERVICE_RESOURCE,
+  SPAN_ID,
+  SPAN_SUBTYPE,
+  SPAN_TYPE,
+} from '../../../../common/elasticsearch_fieldnames';
+import { ProcessorEvent } from '../../../../common/processor_event';
+import { rangeQuery } from '../../../../../observability/server';
+import { Setup } from '../../helpers/setup_request';
+import { withApmSpan } from '../../../utils/with_apm_span';
+import { Node, NodeType } from '../../../../common/connections';
+
+type Destination = {
+  backendName: string;
+  spanId: string;
+  spanType: string;
+  spanSubtype: string;
+} & (
+  | {}
+  | {
+      serviceName: string;
+      agentName: AgentName;
+      environment: string;
+    }
+);
+
+// This operation tries to find a service for a backend, by:
+// - getting a span for each value of span.destination.service.resource (which indicates an outgoing call)
+// - for each span, find the transaction it creates
+// - if there is a transaction, match the backend name (span.destination.service.resource) to a service
+export const getDestinationMap = ({
+  setup,
+  start,
+  end,
+  filter,
+  offset,
+}: {
+  setup: Setup;
+  start: number;
+  end: number;
+  filter: QueryDslQueryContainer[];
+  offset?: string;
+}) => {
+  return withApmSpan('get_destination_map', async () => {
+    const { apmEventClient } = setup;
+
+    const offsetInMs = getOffsetInMs(start, offset);
```

Comment:
Have you considered returning `startWithOffset` and `endWithOffset` in `getOffsetInMs`? I feel like everywhere we'll use it we'll have to subtract the offset from start and end.

```suggestion
    const {offsetInMs, startWithOffset, endWithOffset} = getOffsetInMs(start, offset);
```

Reply:
👍 good idea